Feature/dropdown

youdaeng2 opened the pull request:
# 📦 Pull Request

## 📝 요약(Summary)

- **MenuDropdown**: 수정/삭제 등 액션용 메뉴 드롭다운  
- **SelectDropdown**: 폼 입력용 선택 드롭다운  
- Radix UI 기반으로 구조 설계 및 기본 스타일링 적용  
- 사용 예시 가이드 페이지 추가: `src/pages/guide/dropdown.tsx`
- 뷰포트에 따라 크기와 폰트 사이즈가 달라집니다.

## 💬 공유사항 to 리뷰어


- 처음에 공통 드롭다운을 생성했다가 둘의 기능이 달라 아예 분리하였습니다. 이렇게 하는 거 괜찮을까요?  (둘이 좀 달라서 둘 다 보셔야 할 것 같아용..ㅠ)
- 키보드 포커스는 기본으로 제공되는데, 혹시 더 추가할 부분이 있다면 말씀해주세요.
- focus ring은 디자인적으로.. 마음에 안 들어서 넣지 않았습니다. 추가하는 게 나을 것 같다면 말씀해주세요!

## 🗂️ 관련 이슈

`close #10`

## 📸 스크린샷

<img width="151" height="157" alt="image" src="https://github.com/user-attachments/assets/378c9fe5-e059-4f3b-b711-984a196110e0" />
<img width="121" height="148" alt="image" src="https://github.com/user-attachments/assets/3b9e8d4a-b46a-4f9f-b6b1-b3af71f8fc5f" />
<img width="313" height="186" alt="image" src="https://github.com/user-attachments/assets/b8fb6a17-5dfe-47a6-ab38-78edacf1a7af" />
<img width="423" height="196" alt="image" src="https://github.com/user-attachments/assets/941b09a2-9fe0-4034-bdef-07f878c496e5" />


## ✅ 체크리스트

- [ ] 빌드 및 테스트 통과  
- [ ] ESLint/Prettier 검사 통과

youdaeng2 committed:
feat: 메뉴 드롭다운 생성

youdaeng2 committed:
feat: select 전용 드롭다운 생성

youdaeng2 committed:
feat: 드롭다운 가이드 페이지 생성

youdaeng2 committed:
Merge branch 'dev' of https://github.com/codeit-part3-7/WHYNE_FE into feature/dropdown

youdaeng2 committed:
fix: 파일 경로 수정

youdaeng2 committed:
refactor: SelectDropdown 컴포넌트에 cn 유틸 적용

youdaeng2 committed:
docs: 드롭다운 컴포넌트 Trigger 커스터마이징 주석 추가

626-ju commented on src/components/common/dropdown/MenuDropdown.tsx:
```diff
@@ -0,0 +1,65 @@
+interface MenuDropdownProps {
+  options: Option[]; // 표시할 메뉴 항목들
+  onSelect: (value: string) => void; // 항목 클릭 시 실행할 콜백
+  label?: string; // 버튼 텍스트 (기본: "메뉴")
```
label:string -> label:React.ReactNode 
로 바꿔주시면 제 헤더에 들어가는 유저 이미지도 공통 드랍다운을 사용할 수 있을 것 같습니다!

626-ju commented on src/components/common/dropdown/MenuDropdown.tsx:
```diff
@@ -0,0 +1,65 @@
+      <DropdownMenuTrigger asChild>
+        <button
+          type="button"
+          className="bg-violet-600 text-white px-4 py-2 rounded-lg font-medium  hover:bg-violet-700"
```
{/* 드롭다운 트리거 버튼 - 사용 시 Trigger 컴포넌트를 바꿔 넣어야합니다!*/}
-> 이거가 어떤 의미인지 제가 잘 이해하지 못했습니다.
-> 이게 어떤 의미인지 월요일 회의 때 좀 더 설명해주실 수 있을까요??
-> 상황에 따라서 아래 코멘트는 무시하셔도 될 것 같습니다! :)

이게 지금 MenuDropdown 안에 트리거가 묶여있으면
저희가  호출해서 사용하려고 할 때 
트리거를 커스터마이징 하기 어려울 것 같다는 생각이 들었습니다.
label을 지금 텍스트로 주고 계신데
위에서 차라리 ReactNode타입을 받아서
커스텀한 컴포넌트를 내려주면 재사용성이 더 좋아질 것 같습니다!

-> 만약 이렇게 하게 된다면
button은 아예 없애버리는 대신 
```
<div
 role="button"
 aria-label="드랍다운 버튼"
 //이것 외에도 tabIndex 같은 다른 접근성이 필요하다면 추가하셔도 될 것 같습니다
 >
```
 이런 식으로 대체를 하는 게 좋을 것 같은데  어떻게 생각하시나요?

youdaeng2 replied:
제가 설명이 부족했네요!
기본적으로는 `<DropdownMenuTrigger asChild>` 안에 `<Button>` 등 원하는 트리거 요소를 넣는 게 맞고, 제가 임의로 버튼을 박아뒀던 건 재사용성을 떨어뜨리는 구조가 맞네요.. 생각을 ..못했네요.
트리거 요소가 컴포넌트 안에 있으면 매번 수정을 해줘야한다는 걸 생각을 못했네요. 짚어주셔서 감사합니다

피드백 주신 대로 `trigger`라는 `ReactNode` 프로퍼티를 받아서 그대로 `<DropdownMenuTrigger asChild>`에 넣어주는 방식으로 수정해보겠습니다. 감사합니다!

626-ju replied:
저야말로 감사합니다~

llmojoll reviewed (approved):
고생하셨습니다. 깔끔하게 잘 구현해주신것 같아요 컴포넌트 두개로 분리해서 구현 해주셔서 사용할 때 역할에 맞게 사용할 수 있을 것 같습니다.
ring은 넣지 않는게 디자인적으로 깔끔할 것 같습니다.

626-ju commented on src/components/common/dropdown/MenuDropdown.tsx:
```diff
@@ -0,0 +1,65 @@
+
+      {/* 드롭다운 메뉴 영역 */}
+      <DropdownMenuContent
+        align="start"
```
시안 상으로 MenuDropdown.tsx(케밥모양으로 된거) -> align="end" 인 것 같은데 
확인 한번 부탁드립니다!(아닐 수도 있어요!)

626-ju reviewed (commented):
고생하셨습니다! :)

그 드랍다운 트리거 관련해서만 월요일에 이야기 나눠보면 좋을 것 같습니다!
제가 잘 이해를 못해서요 ㅠ

youdaeng2 committed:
fix: 드롭다운 메뉴 align을 end로 변경

youdaeng2 committed:
refactor: Menudropdown trigger prop 적용 및 내부 버튼 제거

- MenuDropdownProps에 trigger: React.ReactElement prop 추가
- DropdownMenuTrigger asChild에 trigger 요소 적용
- 내부에 하드코딩된 버튼 컴포넌트 제거
- ariaLabel 따로 받을 수 있게 분기

youdaeng2 committed:
refactor: SelectDropdown에 trigger prop 적용 및 선택 라벨 자동 삽입

- SelectDropdownProps에 trigger: React.ReactElement prop 추가
- DropdownMenuTrigger asChild에 trigger 요소를 cloneElement로 적용
- 선택된 항목(label)을 trigger의 children으로 자동 주입
- 사용자 측에서 버튼 텍스트를 별도로 넣지 않아도 되도록 개선

youdaeng2 committed:
Merge branch 'dev' of https://github.com/codeit-part3-7/WHYNE_FE into feature/dropdown

youdaeng2 committed:
docs: DropdownGuidePage에서 trigger prop 구조에 맞게 예제 수정

youdaeng2 committed:
chore: MenuDropdown trigger 사용법 주석 수정

dwkim101 commented on src/components/common/dropdown/MenuDropdown.tsx:
```diff
@@ -0,0 +1,58 @@
+ * - trigger prop으로 원하는 요소를 넘겨서 사용
+ * - 옵션 label에 ReactNode 허용 (이미지, 아이콘 등)
+ */
+interface MenuDropdownProps {
+  options: Option[]; // 표시할 메뉴 항목들
+  onSelect: (value: string) => void; // 항목 클릭 시 실행할 콜백
+  trigger: React.ReactElement;
+}
```
`onSelect`콜백의 파라미터 value는 아래 사용하는 곳을 보니 Option['value'] 타입이군요. 제네릭을 쓰면 더 타입안전한 컴포넌트가 될 수 있겠네요! TS역량을 어필할 수도 있고요 (제네릭은  ts코드를 - 특히 라이브러리나 기존의 코드 - 이해하는데 필수입니다)

```ts
onClick={() => onSelect(option.value)}
```

```suggestion

interface Option<T> {
  label: React.ReactNode;
  value: T;
  ariaLabel?: string;
}

interface MenuDropdownProps<T> {
  options: Option<T>[]; // 표시할 메뉴 항목들
  onSelect: (value: T) => void; // 항목 클릭 시 실행할 콜백
  trigger: React.ReactElement;
}

export default function MenuDropdown<T extends string>({
  options,
  onSelect,
  trigger,
}: MenuDropdownProps<T>) {
```

dwkim101 commented on src/components/common/dropdown/MenuDropdown.tsx:
```diff
@@ -0,0 +1,58 @@
+interface Option {
+  label: React.ReactNode;
+  value: string;
+  ariaLabel?: string;
```
👍

dwkim101 commented on src/components/common/dropdown/MenuDropdown.tsx:
```diff
@@ -0,0 +1,58 @@
+      >
+        {options.map((option) => (
+          <DropdownMenuItem
+            aria-label={option.ariaLabel}
```
옵셔널에는 언제나 fallback값을 고려하는 습관을 두시는게 좋습니다
ariaLabel은 옵셔널이니까 아래처럼 제안드려요
```suggestion
            aria-label={option.ariaLabel ?? option.value}
```
